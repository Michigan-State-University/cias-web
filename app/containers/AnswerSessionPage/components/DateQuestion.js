import React from 'react';
import PropTypes from 'prop-types';

import isNullOrUndefined from 'utils/isNullOrUndefined';

import DateQuestionLayout from '../layouts/DateQuestionLayout';

const DateQuestion = ({
  question,
  answerBody,
  selectAnswer,
  formatMessage,
}) => {
  const {
    body: {
      variable: { name },
    },
  } = question;

  const onChange = (event) => {
    if (isNullOrUndefined(event)) return;
    const date = new Date(event);
    selectAnswer([
      {
        var: name,
        value: date.toDateString(),
      },
    ]);
  };

  return (
    <DateQuestionLayout
      formatMessage={formatMessage}
      onChange={onChange}
      answerBody={answerBody?.[0]}
    />
  );
};

DateQuestion.propTypes = {
  question: PropTypes.object.isRequired,
  selectAnswer: PropTypes.func,
  answerBody: PropTypes.any,
  formatMessage: PropTypes.func,
};

export default DateQuestion;
