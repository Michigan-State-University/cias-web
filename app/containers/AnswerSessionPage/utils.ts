import dayjs from 'dayjs';
import pickBy from 'lodash/pickBy';
import isNil from 'lodash/isNil';

import { TlfbConfigBody } from 'models/Question';
import { CalendarData } from 'models/Tlfb';
import { CamelToSnake } from 'global/types/camelToSnake';
import { fullDayToYearFormatter } from 'utils/formatters';

export const getSessionMapUserPreviewUrl = (userSessionId: string): string => {
  const {
    location: { pathname },
  } = window;
  const previewIndex = pathname.indexOf('preview');
  const sessionUrl = pathname.slice(0, previewIndex - 1);
  return `${sessionUrl}/map?userSessionId=${userSessionId}`;
};

export const getNextSessionUrl = (sessionId: string): string => {
  const {
    location: { pathname },
  } = window;

  return pathname.replace(/\/sessions\/.*/, `/sessions/${sessionId}/fill`);
};

export const getBackToModulesUrl = (): string => {
  const {
    location: { pathname },
  } = window;

  return pathname.replace(/\/sessions\/.*/, '/invite');
};

const generateGridCellId = (prefix: string, index: number) =>
  `${prefix}-${index}`;

export const generateGridAnswerId = (index: number) =>
  `${generateGridCellId('answer', index)}`;
export const generateGridQuestionId = (index: number) =>
  `${generateGridCellId('question', index)}`;

const getOldestDate = (dates: CalendarData) =>
  dayjs.min(
    Object.keys(dates).map((dayString) =>
      dayjs(dayString, fullDayToYearFormatter),
    ),
  );

export const getCalendarMetadata = (
  tlfbConfig: CamelToSnake<TlfbConfigBody>,
  calendarData: CalendarData,
) => {
  const {
    data: [
      {
        payload: { days_count: daysCount },
      },
    ],
  } = tlfbConfig;
  const yesterday = dayjs().subtract(1, 'day');
  const oldestFilledDate = getOldestDate(calendarData);
  const datesWithAnswers = pickBy(calendarData, (date) => !isNil(date.answer));
  const oldestFilledAnswerDate = getOldestDate(datesWithAnswers);
  const oldestAllowedDate = dayjs().subtract(+(daysCount || 1), 'day');
  const isEveryAnswerFilled =
    oldestFilledAnswerDate?.isSame(oldestAllowedDate, 'day') ?? false;

  return {
    yesterday,
    oldestAllowedDate,
    oldestFilledDate,
    oldestFilledAnswerDate,
    isEveryAnswerFilled,
  };
};
