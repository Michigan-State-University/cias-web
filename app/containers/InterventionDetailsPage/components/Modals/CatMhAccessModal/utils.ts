import { Intervention } from 'models/Intervention';

import { ModalUIData } from './types';

export const getTestsLeft = (
  catMhPool: Nullable<number>,
  createdCatMhSessionCount: Nullable<number>,
) => Math.max(0, (catMhPool ?? 0) - (createdCatMhSessionCount ?? 0));

export const mapInterventionToModalData = (
  intervention: Intervention,
): ModalUIData => {
  const {
    catMhPool,
    licenseType,
    catMhApplicationId,
    catMhOrganizationId,
    isAccessRevoked,
    createdCatMhSessionCount,
  } = intervention;

  return {
    testNumber: catMhPool ?? 0,
    testsLeft: getTestsLeft(catMhPool, createdCatMhSessionCount),
    licenseType,
    applicationId: catMhApplicationId ?? '',
    organizationId: catMhOrganizationId ?? undefined,
    isAccessRevoked,
  };
};

export const mapModalDataToIntervention = (
  modalData: ModalUIData,
): Partial<Intervention> => {
  const {
    testNumber,
    licenseType,
    applicationId,
    organizationId,
    isAccessRevoked,
  } = modalData;

  return {
    catMhPool: testNumber ?? 0,
    licenseType,
    catMhApplicationId: applicationId,
    catMhOrganizationId: organizationId,
    isAccessRevoked,
  };
};
