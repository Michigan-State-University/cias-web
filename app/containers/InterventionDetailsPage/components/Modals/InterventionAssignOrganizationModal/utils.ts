import { Organization } from 'models/Organization';

import { SelectOption } from 'components/Select/types';

export const organizationSelectOptionFormatter = ({
  id: value,
  name: label,
}: Organization): SelectOption<string> => ({
  value,
  label,
});
