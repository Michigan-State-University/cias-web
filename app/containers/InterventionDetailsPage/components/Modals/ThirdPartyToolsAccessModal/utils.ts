import { InterventionDto } from 'models/Intervention';

import { ModalUIData } from './types';

export const mapInterventionToModalData = (
  intervention: InterventionDto,
): ModalUIData => {
  const {
    catMhPool,
    licenseType,
    catMhApplicationId,
    catMhOrganizationId,
    isAccessRevoked,
    createdCatMhSessionCount,
    hfhsAccess,
  } = intervention;

  return {
    testNumber: catMhPool ?? 0,
    currentTestNumber: catMhPool - createdCatMhSessionCount ?? 0,
    licenseType,
    applicationId: catMhApplicationId ?? '',
    organizationId: catMhOrganizationId ?? undefined,
    isAccessRevoked,
    hfhsAccess,
  };
};

export const mapModalDataToIntervention = (
  modalData: ModalUIData,
): Partial<InterventionDto> => {
  const {
    testNumber,
    licenseType,
    applicationId,
    organizationId,
    isAccessRevoked,
    hfhsAccess,
  } = modalData;

  return {
    catMhPool: testNumber ?? 0,
    licenseType,
    catMhApplicationId: applicationId,
    catMhOrganizationId: organizationId,
    isAccessRevoked,
    hfhsAccess,
  };
};
