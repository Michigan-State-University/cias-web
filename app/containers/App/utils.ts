import { matchPath } from 'react-router';

import { PasswordAuthenticatedRoles, Roles } from 'models/User/RolesManager';

import { arraysOverlap } from 'utils/arrayUtils';

import { ALL_ROUTE_PATHS, RoutePath } from 'global/constants';

import { OMIT_FETCH_SELF_DETAILS_PATHS } from './constants';

export const shouldFetchSelfDetailsOnPath = (pathname: string) => {
  const match = matchPath(pathname, { path: ALL_ROUTE_PATHS, exact: true });
  return (
    match && !OMIT_FETCH_SELF_DETAILS_PATHS.includes(match.path as RoutePath)
  );
};

export const shouldFetchSelfDetailsByUserRoles = (userRoles: Roles[]) =>
  arraysOverlap(userRoles, PasswordAuthenticatedRoles);
