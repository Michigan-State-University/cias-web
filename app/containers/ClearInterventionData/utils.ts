import * as Yup from 'yup';
import { nonNegativeIntegerValidationSchema } from 'utils/validators';

export const createClearConversationDataFormValidationSchema = () =>
  Yup.object().shape({
    delay: nonNegativeIntegerValidationSchema,
  });
