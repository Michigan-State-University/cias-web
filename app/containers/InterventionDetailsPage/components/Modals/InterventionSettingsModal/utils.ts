import * as Yup from 'yup';

import { Intervention, InterventionType } from 'models/Intervention';
import { ShortLinkData } from 'models/ShortLink';

import { jsonApiToArray } from 'utils/jsonApiMapper';
import { LanguageSelectOption } from 'utils/formatters';
import {
  requiredValidationSchema,
  unreservedURLCharactersSchema,
} from 'utils/validators';

import {
  InterventionSettingsFormValues,
  GetShortLinksResponse,
  ShortLinksData,
} from './types';

// TODO validate duplicates
export const InterventionSettingsFormValidationSchema = Yup.object().shape({
  links: Yup.array().of(
    Yup.object({
      selected: Yup.boolean(),
      name: Yup.string().when('selected', {
        is: (selected) => selected,
        then: unreservedURLCharactersSchema.concat(requiredValidationSchema),
      }),
      healthClinicId: Yup.string(),
    }),
  ),
});

export const getShortLinksDataParser = (
  data: GetShortLinksResponse,
): ShortLinksData => ({
  shortLinks: jsonApiToArray(data, 'shortLink'),
  healthClinics: jsonApiToArray(
    { data: data.health_clinics },
    'simpleHealthClinic',
  ),
});

export const mapShortLinksToFormValues = (
  shortLinksData: Nullable<ShortLinksData>,
  inOrganization: boolean,
): InterventionSettingsFormValues['links'] => {
  if (!inOrganization) {
    const links = shortLinksData?.shortLinks?.map(({ name }) => ({
      selected: !!name,
      name: name ?? '',
    }));
    return links?.length ? links : [{ selected: false, name: '' }];
  }

  return (
    shortLinksData?.healthClinics?.map(({ id }) => {
      const { name } =
        shortLinksData?.shortLinks?.find(
          ({ healthClinicId }) => healthClinicId === id,
        ) ?? {};
      return {
        selected: !!name,
        name: name ?? '',
        healthClinicId: id,
      };
    }) ?? []
  );
};

export const mapFormValuesToShortLinks = (
  links: InterventionSettingsFormValues['links'],
): ShortLinkData[] =>
  links
    .filter(({ selected, name }) => selected && name)
    .map(({ name, healthClinicId }) => ({ name: name!, healthClinicId }));

export const mapLanguageToInterventionChanges = ({
  value,
  label,
  googleLanguageId,
}: LanguageSelectOption): Pick<
  Intervention,
  'languageCode' | 'languageName' | 'googleLanguageId'
> => ({
  languageCode: value,
  languageName: label,
  googleLanguageId: +googleLanguageId,
});

export const getPlaceholderBase = (
  interventionId: string,
  interventionType: InterventionType,
  firstSessionId: Nullable<string>,
) => {
  const base = `${process.env.WEB_URL}/interventions/${interventionId}`;

  switch (interventionType) {
    case InterventionType.DEFAULT:
      if (!firstSessionId) return '';
      return `${base}/sessions/${firstSessionId}/fill`;
    default:
      return `${base}/invite`;
  }
};
