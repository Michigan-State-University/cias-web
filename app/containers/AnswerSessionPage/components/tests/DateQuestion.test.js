/**
 *
 * Tests for DateQuestion
 *
 */

import React from 'react';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { DEFAULT_LOCALE } from 'i18n';

import { formatMessage } from 'utils/intlOutsideReact';
import { createTestStore } from 'utils/testUtils/storeUtils';

import DateQuestion from '../DateQuestion';

describe('<DateQuestion />', () => {
  const mockedFunctions = {
    selectAnswer: jest.fn(),
    formatMessage,
    showError: jest.fn(),
  };

  let store;
  beforeAll(() => {
    store = createTestStore();
  });

  const defaultProps = {
    question: {
      body: {
        variable: 'test',
      },
    },
    answerBody: [],
    ...mockedFunctions,
  };

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <MemoryRouter>
            <DateQuestion {...defaultProps} />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const { container } = render(
      <Provider store={store}>
        <IntlProvider locale={DEFAULT_LOCALE}>
          <MemoryRouter>
            <DateQuestion {...defaultProps} />
          </MemoryRouter>
        </IntlProvider>
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });
});
