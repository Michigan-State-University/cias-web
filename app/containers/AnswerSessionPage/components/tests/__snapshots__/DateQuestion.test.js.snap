// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateQuestion /> Should render and match the snapshot 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  height: auto;
  width: 100%;
}

.c6 {
  padding: 12px;
  border-style: solid;
  border-width: 1px;
  border-color: #E9F1FB;
  border-radius: 5px;
  margin-right: 10px;
  margin-left: 10px;
  width: 200px;
  height: 50px;
  padding: 12px;
  font-size: 15px;
  color: rgba(47,56,80,1);
  text-align: left;
  color: rgba(47,56,80,1);
  border: 3px solid #107969;
}

.c6:focus {
  box-shadow: none;
  outline: none;
  border-color: #107969;
}

.c6:disabled {
  color: #ABBBD1;
}

.c5 .schedule-date-picker {
  font-family: DM Sans;
  background-color: #E9F1FB;
}

.c5 .schedule-date-picker .react-datepicker__header {
  background-color: #F8FBFF;
}

.c5 .schedule-date-picker .react-datepicker__day {
  color: #2F3850;
}

.c5 .schedule-date-picker .react-datepicker__day:hover {
  color: #FFFFFF;
  background-color: rgba(200,102,234,0.7);
}

.c5 .schedule-date-picker .react-datepicker__day--keyboard-selected {
  background-color: #C866EA;
}

.c5 .schedule-date-picker .react-datepicker__day--in-range,
.c5 .schedule-date-picker .react-datepicker__day--in-selecting-range {
  color: #2F3850;
  background-color: rgba(200,102,234,0.1);
}

.c4 .schedule-date-picker .react-datepicker__day--selected:not(.c4 .schedule-date-picker .react-datepicker__day--in-selecting-range),
.c5 .schedule-date-picker .react-datepicker__day--range-start:not(.c5 .schedule-date-picker .react-datepicker__day--in-selecting-range),
.c5 .schedule-date-picker .react-datepicker__day--range-end:not(.c5 .schedule-date-picker .react-datepicker__day--in-selecting-range),
.c5 .schedule-date-picker .react-datepicker__day--selecting-range-start,
.c5 .schedule-date-picker .react-datepicker__day--selecting-range-end {
  color: #FFFFFF;
  background-color: #C866EA;
}

.c5 .schedule-date-picker .react-datepicker__year-option {
  background-color: #E9F1FB;
}

.c5 .schedule-date-picker .react-datepicker__month-option {
  background-color: #E9F1FB;
}

.c5 .schedule-date-picker .react-datepicker__time-list-item:hover {
  color: #FFFFFF;
  background-color: rgba(200,102,234,0.7) !important;
}

.c5 .schedule-date-picker .react-datepicker__time-list-item--selected {
  background-color: #C866EA !important;
}

.c5 .react-datepicker-wrapper {
  width: 100%;
}

.c0 {
  width: auto;
  height: auto;
  display: block;
  border-radius: 5px;
  width: 100%;
  padding: 15px;
  color: unset;
}

@media only screen and (min-width:768px) {

}

@media only screen and (min-width:960px) {

}

@media only screen and (min-width:1170px) {

}

@media only screen and (min-width:1360px) {

}

<div>
  <div
    class="c0"
    width="100%"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        width="100%"
      >
        <div
          class="c3"
        >
          <div
            class="c4 c5"
          >
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <input
                  aria-label="MM/DD/YYYY"
                  class="c6"
                  color="#2F3850"
                  dir="auto"
                  font-size="15"
                  height="50"
                  inputmode="numeric"
                  placeholder="MM/DD/YYYY"
                  type="text"
                  value=""
                  width="200"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
